Topics/Tech_Stacks/Spring-Security.md

```diff
@@ -0,0 +1,218 @@
+# REST API Security
+- Spring Security Model
+	- Spring Security defines a framework for security
+	- Implemented using Servlet filters in the background
+		- Servlet filters pre-process / post-process web requests
+		- Servlet filters route web requests based on security logic
+		- Spring provides a bulk of security functionality with servlet filters
+	- Two methods of securing an app: declarative and programmatic
+
+### Security Concepts
+- Authentication
+	- Check user id and password with credentials stored in app / db
+- Authorization
+	- Check to see if user has an authorized role
+- Declarative Security
+	- Define application's security constraints in configuration
+		- All Java config: @Configuration
+	- Provides separation of concerns between application code and security
+- Programmatic Security
+	- Spring Security provides an API for custom application coding
+	- Provides greater customization for specific app requirements
+
+### Enabling Spring Security
+1. Edit pom.xml and add `spring-boot-starter-security`
+2. This will automagically secure all endpoints for application
+	- Accessing secured endpoints will prompt for login
+	- Override default user name and generated password in application.properties
+
+``` Java
+// Spring Security Configuration
+@Configuration
+@EnableWebSecurity
+public class SecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        http
+            .authorizeRequests()
+            .anyRequest().authenticated()
+            .and()
+            .formLogin()
+            .and()
+            .httpBasic();
+    }
+}
+```
+
+- Dev Process
```

Comment:
The addition of the practical development process greatly enhances the content, making it easier for me to grasp the concept and its application.